oxenss/snode/service_node.cpp

```diff
@@ -648,48 +738,46 @@ void ServiceNode::update_swarms() {
                         log::info(logcat, "Got initial swarm information from local Oxend");
-
-                        {
-                            std::lock_guard l{first_response_mutex_};
-                            got_first_response_ = true;
-                        }
-                        first_response_cv_.notify_all();
-
-                        // Request some recent block hash heights so that we can properly carry out
-                        // and respond to storage testing (for which we need to know recent block
-                        // hashes). Incoming tests are *usually* height - TEST_BLOCKS_BUFFER, but
-                        // request a couple extra as a buffer.
-                        for (uint64_t h = bu.height - TEST_BLOCKS_BUFFER - 2; h < bu.height; h++)
-                            omq_server_.oxend_request(
-                                    "rpc.get_block_hash",
-                                    [this, h](bool success, std::vector<std::string> data) {
-                                        if (!(success && data.size() == 2 && data[0] == "200" &&
-                                              data[1].size() == 66 && data[1].front() == '"' &&
-                                              data[1].back() == '"'))
-                                            return;
-                                        std::string_view hash{
-                                                data[1].data() + 1, data[1].size() - 2};
-                                        if (oxenc::is_hex(hash)) {
-                                            log::debug(
-                                                    logcat,
-                                                    "Pre-loaded hash {} for height {}",
-                                                    hash,
-                                                    h);
-                                            block_hashes_cache_.insert_or_assign(h, hash);
-                                        }
-                                    },
-                                    "{\"height\":[" + util::int_to_string(h) + "]}");
+                        static_assert(
+                                server::OMQ::NUM_GENERAL_THREADS == 1,
+                                "If the number of threads ever increases from 1, there's a race "
+                                "condition here with `on_swarm_update` both writing to "
+                                "`block_hashes_cache` so writes will need to be synchronised. "
+                                ""
+                                "Since OMQ was integrated in 330581b this has always been set to 1 "
+                                "so all these 'async' OMQ requests that SS sends is actually "
+                                "thread safe even though a typical configuration of OMQ (w/ "
+                                "threads scaling with the number of cores) would trigger races.");
```

Comment:
This description is very far from true, and so if there is such a race condition it needs to be solved.  In storage server's use of OxenMQ there can be up to 6 worker threads running at once in the current configuration:
- `sn.*` commands (invocable by other SNs) have 2 reserved threads
- `storage.*` (invoked by clients) has one reserved thread
- `monitor.*` (invoked for push notifications) has one reserved thread
- `server.*`, `notify.*`, and timers fall under the single general thread
- reply processing from outgoing requests have one reserved thread.

Reply:
This is removed now with the removed storage testing. The startup sequence with the hashes had 2 or 3 races alone but since that's been unused for 2+ years then I can see how that's just been sitting there and that it also happened to be the first part of the codebase I started investigating.